[Streams 🌊] Prevent stream partition names from messing up required hierarchy prefix

AlejandroFrndz opened the pull request:
Closes [#576](https://github.com/elastic/streams-program/issues/576)

## Summary

When partitioning a wired stream, each partition has restrictions in their name by which they must keep the name of their parent partition(s) keeping the hierarchy separated by dots in each level.

While server-side the restriction is properly enforced, the input we're rendering at UI level does not enforce this restriction. Instead it allows users to alter and/or remove the required prefix. Even if an error message will be presented should the name not meet the requirements, it still isn't an ideal UX given the input does allow you to alter the prefix without clearly stating it shouldn't be modified.

This PR introduces some fixes to address the issue

## UI
### Partition Creation
During creation, the stream name input has now been split into 2 sections. The fixed prefix part of the partition name is now rendered in a prepend section for the input, making it non editable. The child part of the name remains as a regular text input allowing users to name the new partition however they wish to, but keeping the prefix restrictions.


https://github.com/user-attachments/assets/cbf6df0f-47ee-4959-ad6b-a77da78605a4


### Partition Update
During update, editing the name is not allowed. However, to keep visual consistency, the prefix will still be separated and rendered as a prepend. The input will not be editable, but it has been updated from using `disabled` into using `readOnly`.

<img width="1000" height="411" alt="Screenshot 2025-10-29 at 16 20 31" src="https://github.com/user-attachments/assets/660f22fd-550c-463c-b9a4-d5b219ce361c" />

## Gotchas
### Prefix Truncation
Given the nesting nature of partitions, the prefix can become too long to be fully render while keeping enough space for the partition name input. Generally, we try to render at most 25 prefix characters. Anything above that will be truncated. In the case of narrow screens, the prefix prepend will never be larger than half the total size of the stream name input, regardless of characters, so anything that doesn't fit in this limitation will also be truncated.

<img width="993" height="416" alt="Screenshot 2025-10-29 at 16 26 14" src="https://github.com/user-attachments/assets/2a4d910a-7fd2-4614-ab40-b6ae2e05d56f" />

### Invalid partition names
As we've seen before, each partition is composed of a prefix with the name of it's ancestor streams plus the partition name itself. The way we handle nesting in the prefix is by separating each partition name from it's predecessors with a dot. These dots are added automatically by the system when creating partitions and the hierarchy *must* be kept in the name, otherwise the backend will reject the partition creation.

At UI level, we didn't enforce this restriction before which could lead users to attempt to create a partition with dots on the name, only to be met with an error description that can be confusing.

<img width="333" height="208" alt="Screenshot 2025-10-30 at 17 02 05" src="https://github.com/user-attachments/assets/8382b474-ed68-4ca0-b778-79799b2b55af" />

Notice how the error states that the parent stream and a dot must be included, which the name `log.child.child` name does satisfy. At no point is it explicit that partition names cannot contain dots

Instead on relaying on the server side error, this PR introduces validation at UI level in order to display a user friendly error message letting them know that dots cannot be used in partition names.

<img width="993" height="431" alt="Screenshot 2025-10-30 at 16 54 28" src="https://github.com/user-attachments/assets/c9ba3144-fc9f-4b6c-b5f1-3ec1aa103b2a" />


<!--ONMERGE {"backportTargets":["9.2"]} ONMERGE-->

AlejandroFrndz committed:
feat: Enhance StreamNameFormRow component with prefix handling and accessibility improvements

- Refactored StreamNameFormRow to separate stream prefix and partition name using useMemo.
- Updated input handling to reflect changes in partition name while preserving the prefix.
- Added accessibility features including screen reader support and improved help text.
- Introduced a new constant for maximum visible characters for the prefix label.

AlejandroFrndz committed:
refactor: Update StreamNameFormRow to use readOnly prop instead of disabled

- Changed the disabled prop to readOnly in StreamNameFormRow for better clarity and functionality.
- Updated the EditRoutingStreamEntry component to reflect this change, ensuring the stream name is displayed as read-only.

AlejandroFrndz committed:
test: Update tests to reflect new prefix based partition naming

- Updated routing rule names in various test files to remove the 'logs.' prefix in input content.
- Adjusted related assertions to ensure visibility of updated stream names during routing rule creation and editing.

AlejandroFrndz committed:
fix: Prevent dots in partition names

- Prevent partition creation when the resulting stream doesn't satisfy childOf restrictions
- Updated selectors to include selectCurrentStream for improved state management.
- Implemented validation to prevent the inclusion of '.' in the partition name, enhancing user input feedback with error messages.

AlejandroFrndz committed:
test: Add validation for routing rule creation

- Introduced a new test to ensure that creating a routing rule with a partition name that is not a child of the current stream is disabled.

tonyghiani reviewed (commented):
Overall great job! I left a couple of notes, but nothing critical to change :)

AlejandroFrndz committed:
refactor: Simplify StreamNameFormRow and remove unused selector

- Updated StreamNameFormRow to directly access the stream name from the context definition, simplifying the logic and reducing reselect overhead.
- Refactored the partition name calculation to eliminate the use of useMemo, enhancing readability.
- Removed the selectCurrentStream selector from the state management as it was no longer needed.

AlejandroFrndz committed:
refactor: Update StreamNameFormRow to use array for prepend prop

- Changed the prepend prop in StreamNameFormRow from a JSX fragment to an array format for improved readability and consistency.
- This refactor maintains the existing functionality while enhancing the structure of the component.

tonyghiani reviewed (approved):
LGTM, good job!

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/357110)
* Commit: 8434be7297ad1b19968830433e6d132663053354

### Failed CI Steps
* [FTR Configs #48](https://buildkite.com/elastic/kibana-pull-request/builds/357110#019a39d3-f3d0-472a-a816-8eb1d4c759db)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b663900c2359de6b9bc9653c7a0c4d7e5e1c1d5e) | [after](https://github.com/elastic/kibana/commit/8434be7297ad1b19968830433e6d132663053354) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 907.6KB | 908.5KB | +865.0B |



### History
* :green_heart: [Build #356867](https://buildkite.com/elastic/kibana-pull-request/builds/356867) succeeded bf14a4233b59cf079927b1e382b2046657dd7a1d
* :green_heart: [Build #356315](https://buildkite.com/elastic/kibana-pull-request/builds/356315) succeeded 0a3042fac97f982b0d6a6818fa91396d52287b23

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357110","number":357110,"commit":"8434be7297ad1b19968830433e6d132663053354"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356867","number":356867,"commit":"bf14a4233b59cf079927b1e382b2046657dd7a1d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356315","number":356315,"commit":"0a3042fac97f982b0d6a6818fa91396d52287b23"}],"number":357110}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 9.2

https://github.com/elastic/kibana/actions/runs/18971938504

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/241447">](https://github.com/elastic/kibana/pull/241447)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
